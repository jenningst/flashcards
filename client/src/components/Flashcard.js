import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { useKeyPress } from '../hooks/use-key-press';

import { Body } from './elements/Text';
import { PrimaryButton } from './elements/Button';

function Flashcard({ card }) {
  const [showAnswer, setShowAnswer] = useState(false);
  const toggleAnswer = () => setShowAnswer(!showAnswer);
  const spacebarPressed = useKeyPress(32);

  useEffect(() => {
    if (spacebarPressed) {
      console.log('down');
      setShowAnswer(!showAnswer);
    }
  }, [spacebarPressed]);

  return (
    <FlashcardWrapper className="Flashcard">
      <ContentArea className="Flashcard__content ContentArea">
        <Body
          className="ContentArea__text"
          data-testid='card-text'
        >
          {!showAnswer
            ? card.text
            : card.answer
          }
        </Body>
      </ContentArea>
      <ToggleButton
        className="Flashcard__button-toggle"
        data-testid="card-button"
        type="button"
        onClick={toggleAnswer}
      >
        {!showAnswer ? 'SHOW ANSWER' : 'SHOW QUESTION'}
      </ToggleButton>
    </FlashcardWrapper>
  );
};

const FlashcardWrapper = styled.div`
  box-sizing: border-box;
  flex-grow: 1;
  
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  height: 100%;
  max-width: 500px;
  padding: 1rem;
  
  border-bottom-left-radius: .50rem;
  border-bottom-right-radius: .50rem;
  box-shadow: 0 4px 12px 0 rgba(0, 0, 0, 0.05);
  background: ${props => props.theme.color.background.pureWhite};
  color: ${props => props.theme.color.font.secondary};
`;

const ContentArea = styled.section`
  flex-grow: 1;
  width: 100%;
  margin: 0rem 1.5rem 1.5rem 1.5rem;

  p {
    font-weight: 300;
    font-style: italic;
  }
`;

const ToggleButton = styled(PrimaryButton)`
  min-width: 224px;
  max-width: 500px;
`;

Flashcard.propTypes = {
  card: PropTypes.object.isRequired,
};

export default Flashcard;
