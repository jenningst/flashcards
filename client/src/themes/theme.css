/* 

button: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      active: '#5860A5',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    border: {
      primary: '#42487C',
      active: '#939CE8'
      disabled: '#b1aeb7',
      input: '#e9eaf0',
    }, // TODO: consolidate everything below; using the above
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }

}*/