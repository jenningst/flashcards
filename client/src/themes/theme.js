export const darkTheme = {
  font: {
    primary: '#FFFFFF',
    secondary: '#E0E0E0',
    tertiary: '#666666',
  },
  background: {
    primary: '#282626',
    secondary: '#313131',
    highlight: '#6121BF',
    special: '#7358F7',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};

export const lightTheme = {
  color: {
    button: {
      primary: '#939CE8',
      primaryHover: '#7883E2',
      active: '#5860A5',
      secondary: '#50cdff',
      secondaryHover: '#6FD6FF',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#33364d',
      blackCoral: '#585A6D',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1',
      placeholder: '#e9eaf0',
      danger: '#DE6A60',
    },
    border: {
      primary: '#5860A5',
      secondary: '#7883E2',
      disabled: '#b1aeb7',
      inputDefault: '#e9eaf0',
      transparent: '#0000ffff',
      eerieBlack: '#16171D',
      charleston: '#33364d',
    }, // TODO: consolidate everything below; using the above
    background: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
    },
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }
  },
  spacing: {
    compact: '1rem',
    comfortable: '1.5rem',
    generous: '2rem',
  },
  font: {
    primary: '#4e5167',
    secondary: '#262626',
    tertiary: '#666666',
  },
  background: {
    primary: '#ffffff',
    secondary: '#fcfbfc',
    highlight: '#9388ff',

    special: '#50cdff',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};