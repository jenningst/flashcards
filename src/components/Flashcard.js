import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import ThemeContext from '../contexts/themeContext';

import { Title1, Body } from '../components/Elements/Text';
import { MediumButton } from '../components/Elements/Button';

Flashcard.propTypes = {
  text: PropTypes.string.isRequired,
  isAnswer: PropTypes.bool.isRequired,
  toggle: PropTypes.func.isRequired,
};

function Flashcard({ text, isAnswer, toggle }) {
  const theme = useContext(ThemeContext);
  const FlashcardWrapper = styled.div`
    box-sizing: border-box;
    display: flex;
    flex-flow: column nowrap;
    justify-content: center;
    align-items: center;
    height: 100%;
  
    border-radius: 20px;
    margin-left: 1rem;
    margin-right: 1rem;
    padding: 1.5rem 1rem 1rem 1rem;
    background: ${theme.background.secondary};
    color: ${theme.font.secondary};
  
    .Flashcard__question {
      flex-grow: 1;
      width: 100%;
      
      .Flashcard__question-text {
        margin: 0;
      }
    }
  
    .Flashcard__toggle-button {
      background: ${theme.button.default.primary};
      border: 2px solid ${theme.button.default.primary};
      color: ${theme.font.primary};
      outline: none;

      &:hover {
        background: ${theme.button.hover.primary};
        border: 2px solid ${theme.button.hover.primary};
      }

      &:active {
        border: 2px solid ${theme.button.active.primary};
      }

      border-radius: 15px;
      width: 50%;
    }
  `;

  return (
    <FlashcardWrapper className="Flashcard">
      <div className="Flashcard__question">
        <Body className="Flashcard__question-text">
          {text}
        </Body>
      </div>
      <MediumButton
        className="Flashcard__toggle-button" type="button" onClick={toggle}>
        {"Flip"}
      </MediumButton>
    </FlashcardWrapper>
  );
};

export default Flashcard;
