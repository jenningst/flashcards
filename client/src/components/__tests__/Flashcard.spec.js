import React from 'react';
import { ThemeProvider } from 'styled-components';
import { darkTheme, lightTheme } from '../../themes/theme';

import {
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';

import Flashcard from '../Flashcard';

afterEach(cleanup);

const renderComponent = ({ question, theme }) =>
  render(
    <ThemeProvider theme={theme}>
      <Flashcard question={question} />
    </ThemeProvider>
  );

describe('<Flashcard /> spec', () => {
  it('assert it matches snapshot', () => {
    const mockObject = {
      id: '12',
      text: 'What is 2 + 2?',
      answer: '4',
    };
    const { asFragment } = renderComponent({ question: mockObject, theme: lightTheme});

    expect(asFragment()).toMatchSnapshot();
  });

  it('assert it renders question text', () => {
    const mockObject = {
      id: '12',
      text: 'What is 2 + 2?',
      answer: '4',
    };
    const { getByTestId } = renderComponent({ question: mockObject, theme: lightTheme});

    expect(getByTestId('card-text')).toHaveTextContent('What is 2 + 2?');
  });

  it('assert it initially renders a SHOW ALL button', () => {
    const mockObject = {
      id: '12',
      text: 'What is 2 + 2?',
      answer: '4',
    };
    const { getByTestId } = renderComponent({ question: mockObject, theme: lightTheme});
    const buttonElement = getByTestId('card-button');

    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveTextContent('SHOW ANSWER');
  });

  it('assert it displays different text when the button is clicked', async () => {
    const mockObject = {
      id: '12',
      text: 'What is 2 + 2?',
      answer: '4',
    };
    const { getByTestId, findByTestId } = renderComponent({ question: mockObject, theme: lightTheme});

    expect(getByTestId('card-text')).toHaveTextContent('What is 2 + 2?');
    fireEvent.click(getByTestId('card-button'), { button: 0 });

    expect(await findByTestId('card-text')).toHaveTextContent('4');
  });

  it('assert it displays different text when the button is clicked', async () => {
    const mockObject = {
      id: '12',
      text: 'What is 2 + 2?',
      answer: '4',
    };
    const { getByTestId, findByTestId } = renderComponent({ question: mockObject, theme: lightTheme});
    const buttonElement = getByTestId('card-button');

    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveTextContent('SHOW ANSWER');

    fireEvent.click(buttonElement, { button: 0 });

    expect(await findByTestId('card-button')).toBeInTheDocument();
    expect(await findByTestId('card-button')).toHaveTextContent('SHOW QUESTION');
  });
});