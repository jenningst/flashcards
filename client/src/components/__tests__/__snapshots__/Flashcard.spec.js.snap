// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flashcard /> spec assert it matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Flashcard sc-jTzLTM iGGYsv"
  >
    <section
      class="Flashcard__content ContentArea sc-fjdhpX leZmqx"
    >
      <p
        class="ContentArea__text sc-EHOje gzeaHE"
        data-testid="card-text"
      >
        What is 2 + 2?
      </p>
    </section>
    <button
      class="sc-iwsKbI sc-gZMcBi Flashcard__button-toggle sc-jzJRlG dBVjFL"
      data-testid="card-button"
      type="button"
    >
      SHOW ANSWER
    </button>
  </div>
</DocumentFragment>
`;
